src/porepy/utils/tangential_normal_projection.py

```diff
@@ -193,74 +175,114 @@ def project_normal(self, num=None):
         # Return the restricted matrix
         return remove_tangential_components * full_projection
 
-    def local_projection(self, ind=None):
-        """Get the local projection matrix (reference)
+    ### Helper functions below
+    def _decompose_vector(self, nc: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
+        """Decompose a set of normal vectors into a basis for the tangential space
+        and the normal space.
 
         Parameters:
-            ind (int, optional): Index (referring to the order of the normal vectors
-                provided to __init__) of the basis to return. Defaults to the first one.
+            nc: Normal vectors ``shape=(dim, num_vecs)``.
 
         Returns:
-            np.array (self.dim x self.dim): Local projection matrix. Multiplication
-                gives projection to the tangential space (first self.dim - 1 rows)
-                and normal space (last)
+            Basis for the tangential and normal space, and the normal vector itself.
 
-        """
-        if ind is None:
-            ind = 0
-        return self.projection[:, :, ind]
-
-    ### Helper functions below
+            np.ndarray: ``shape=(dim, dim, num_vecs)``. The first dim-1 columns are
+                the basis for the tangential space, the final column is the normal
+                vector.
 
-    def _decompose_vector(self, nc):
-        if self.dim == 3:
-            t1 = np.random.rand(self.dim, 1) * np.ones(self.num_vecs)
-            t2 = np.random.rand(self.dim, 1) * np.ones(self.num_vecs)
-            normal, tc1, tc2 = self._gram_schmidt(nc, t1, t2)
-            basis = np.hstack([tc1, tc2, normal])
-        else:
-            t1 = np.random.rand(self.dim, 1) * np.ones(self.num_vecs)
-            normal, tc1 = self._gram_schmidt(nc, t1)
-            basis = np.hstack([tc1, normal])
-        return basis, normal
+            np.ndarray: ``shape=(dim, num_vecs)``. The normal vector.
 
-    def _gram_schmidt(self, u1, u2, u3=None):
         """
-        Perform a Gram Schmidt procedure for the vectors u1, u2 and u3 to obtain a set of
-        orthogonal vectors.
 
-        Parameters:
-            u1: ndArray
-            u2: ndArray
-            u3: ndArray
+        if self.dim == 2:
+            # Normalize the normal vector, just to be sure
+            normal = nc / np.linalg.norm(nc, axis=0)
+            # The tanegntial vector in 2d is deterministic up to an arbitrary sign. We
```

Comment:
Great to fix this!
Please document what happens if the normal is [+-1, 0], so that the tangential is [0, +-1], i.e., has no x component.

Reply:
Implemented.